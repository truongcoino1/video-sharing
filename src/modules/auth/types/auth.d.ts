export type User = {
  id: number
  email: string
  password: string
  refreshToken: string;
  token: string;
}

export type UserSession = Omit<User, 'password' | 'refreshToken'>
