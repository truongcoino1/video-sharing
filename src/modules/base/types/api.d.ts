export type ApiResponse<T> = {
  status: string
  message?: string
  result?: T
}
