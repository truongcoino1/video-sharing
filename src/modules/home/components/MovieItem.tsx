import React, { useState, useCallback, memo } from "react";
import VideoPlaceholder from "./VideoPlaceholder";
import type { Movie } from "../types";

type MovieItemProps = {
  movie: Movie;
};
const MovieItem = ({ movie }: MovieItemProps): JSX.Element => {
  const [hasLoadIframe, setHasLoadIframe] = useState(false);

  const handleClickPlaceholder = useCallback(() => {
    setHasLoadIframe(true);
  }, []);

  return (
    <div className="movie-item lg:flex justify-between mb-12">
      {hasLoadIframe ? (
        <div className="lg:w-[350px] lg:h-[196px] rounded-large w-full flex-shrink-0 overflow-hidden relative">
          <iframe 
            className="w-full h-full rounded-large object-cover"
            title={movie.title}
            loading="lazy"
            allow="autoplay; fullscreen; accelerometer; encrypted-media; gyroscope; picture-in-picture"
            src={`https://www.youtube.com/embed/${movie.youtube_id}?autoplay=1`}
          />
        </div>
      ) : (
        <VideoPlaceholder
          thumbnail={movie.thumbnail}
          handleClick={handleClickPlaceholder}
        />
      )}
      <div className="lg:h-[196px] lg:ml-24 flex-grow lg:mt-0 lg:mb-0 mt-12 mb-24">
        <div className="font-bold text-[28px] overflow-hidden line-clamp-2 text-cyan-400">
          <a
            className="title-link"
            href={`https://www.youtube.com/watch?v=${movie.youtube_id}`}
            target="_blank"
            rel="noopener noreferrer"
          >
            {movie.title}
          </a>
        </div>
        <div className="flex items-center">
          <div className="shared-by text-light-low" title={movie.shared_by}>
            <span className="font-semibold text-16 text-light-high">Shared by:</span> {movie.shared_by}
          </div>
        </div>
        <div className="font-semibold text-16 text-light-high">Description:</div>
        <p className="text-14 overflow-hidden text-light-low line-clamp-2 video-description">{movie.description}</p>
      </div>
    </div>
  );
};

export default memo(MovieItem);
