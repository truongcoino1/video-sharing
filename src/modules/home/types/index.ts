export type Movie = {
  id: string;
  title: string;
  thumbnail: string;
  shared_by: string;
  description: string;
  youtube_id: string;
  created_date: number;
};

export type MovieSession = Omit<Movie, "shared_by" | "id" | "created_date">