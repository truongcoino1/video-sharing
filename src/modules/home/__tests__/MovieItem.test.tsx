/**
 * @jest-environment jsdom
 */
import { render, cleanup } from "@testing-library/react";
import MovieItem from "../components/MovieItem";
import { Movie } from "../types";
import { act } from "react-dom/test-utils";

const movie: Movie = {
  id: "123",
  title: "title",
  description: "description",
  thumbnail: "https://i.ytimg.com/vi/bP9gMpl1gyQ/sddefault.jpg",
  shared_by: "shared_by",
  youtube_id: 'dsdss'
};

describe("MovieItem", () => {
  afterEach(cleanup);

  it("MovieItem should render without crash", () => {
    const result = render(<MovieItem movie={movie} />);
    const elm = result.container.querySelector(".movie-item");
    expect(elm).toBeInTheDocument();
  });

  it("MovieItem should render iframe when click on placeholder", () => {
    const result = render(<MovieItem movie={movie} />);
    const placeholder = result.container.querySelector(
      ".video-placeholder"
    ) as HTMLElement;
    act(() => {
      placeholder?.click();
    });
    const iframe = result.container.querySelector("iframe");
    expect(iframe).toBeInTheDocument();
  });

  it("MovieItem should render title with link", () => {
    const result = render(<MovieItem movie={movie} />);
    const title = result.container.querySelector(
      ".title-link"
    ) as HTMLElement;
    expect(title).toBeInTheDocument();
    expect(title.innerHTML).toBe(movie.title);
    expect(title.getAttribute("href")).toBe(
      `https://www.youtube.com/watch?v=${movie.id}`
    );
  });

  it("MovieItem should render description", () => {
    const result = render(<MovieItem movie={movie} />);
    const description = result.container.querySelector(
      ".video-description"
    ) as HTMLElement;
    expect(description).toBeInTheDocument();
    expect(description.innerHTML).toBe(movie.description);
  });

  it("MovieItem should render shared_by", () => {
    const result = render(<MovieItem movie={movie} />);
    const sharedBy = result.container.querySelector(
      ".shared-by"
    ) as HTMLElement;
    expect(sharedBy).toBeInTheDocument();
    expect(sharedBy.innerHTML).toBe(`<span class=\"font-semibold text-16 text-light-high\">Shared by:</span> ${movie.shared_by}`);
  });

  it("should render correct MovieItem", () => {
    const element = render(<MovieItem movie={movie} />);
    expect(element).toMatchSnapshot();
  });
});
