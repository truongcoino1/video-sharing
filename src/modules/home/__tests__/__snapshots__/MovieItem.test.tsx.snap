// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieItem should render correct MovieItem 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="movie-item lg:flex justify-between mb-12"
      >
        <div
          class="video-placeholder cursor-pointer rounded-large flex items-center justify-center overflow-hidden relative w-full lg:w-[350px] min-w-[350px] lg:h-[196px]"
        >
          <img
            alt="video placeholder"
            class="w-full lg:w-[350px] rounded-large lg:h-[196px] object-cover"
            height="480"
            src="https://i.ytimg.com/vi/bP9gMpl1gyQ/sddefault.jpg"
            width="640"
          />
          <img
            alt="button play"
            class="absolute opacity-50"
            height="72"
            src="/youtube-brands.svg"
            width="72"
          />
        </div>
        <div
          class="lg:h-[196px] lg:ml-24 flex-grow lg:mt-0 lg:mb-0 mt-12 mb-24"
        >
          <div
            class="font-bold text-[28px] overflow-hidden line-clamp-2 text-cyan-400"
          >
            <a
              class="title-link"
              href="https://www.youtube.com/watch?v=123"
              rel="noopener noreferrer"
              target="_blank"
            >
              title
            </a>
          </div>
          <div
            class="flex items-center"
          >
            <div
              class="shared-by text-light-low"
              title="shared_by"
            >
              <span
                class="font-semibold text-16 text-light-high"
              >
                Shared by:
              </span>
               
              shared_by
            </div>
          </div>
          <div
            class="font-semibold text-16 text-light-high"
          >
            Description:
          </div>
          <p
            class="text-14 overflow-hidden text-light-low line-clamp-2 video-description"
          >
            description
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="movie-item lg:flex justify-between mb-12"
    >
      <div
        class="video-placeholder cursor-pointer rounded-large flex items-center justify-center overflow-hidden relative w-full lg:w-[350px] min-w-[350px] lg:h-[196px]"
      >
        <img
          alt="video placeholder"
          class="w-full lg:w-[350px] rounded-large lg:h-[196px] object-cover"
          height="480"
          src="https://i.ytimg.com/vi/bP9gMpl1gyQ/sddefault.jpg"
          width="640"
        />
        <img
          alt="button play"
          class="absolute opacity-50"
          height="72"
          src="/youtube-brands.svg"
          width="72"
        />
      </div>
      <div
        class="lg:h-[196px] lg:ml-24 flex-grow lg:mt-0 lg:mb-0 mt-12 mb-24"
      >
        <div
          class="font-bold text-[28px] overflow-hidden line-clamp-2 text-cyan-400"
        >
          <a
            class="title-link"
            href="https://www.youtube.com/watch?v=123"
            rel="noopener noreferrer"
            target="_blank"
          >
            title
          </a>
        </div>
        <div
          class="flex items-center"
        >
          <div
            class="shared-by text-light-low"
            title="shared_by"
          >
            <span
              class="font-semibold text-16 text-light-high"
            >
              Shared by:
            </span>
             
            shared_by
          </div>
        </div>
        <div
          class="font-semibold text-16 text-light-high"
        >
          Description:
        </div>
        <p
          class="text-14 overflow-hidden text-light-low line-clamp-2 video-description"
        >
          description
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
